lib/posix/pthread.c

```diff
@@ -375,30 +530,26 @@ int pthread_once(pthread_once_t *once, void (*init_func)(void))
  *
  * See IEEE 1003.1
  */
+FUNC_NORETURN
 void pthread_exit(void *retval)
 {
-	k_spinlock_key_t cancel_key;
-	struct posix_thread *self = to_posix_thread(pthread_self());
+	sys_snode_t *node_l;
+	k_spinlock_key_t key;
 	pthread_key_obj *key_obj;
+	struct posix_thread *self;
 	pthread_thread_data *thread_spec_data;
-	sys_snode_t *node_l;
 
-	/* Make a thread as cancelable before exiting */
-	cancel_key = k_spin_lock(&self->cancel_lock);
-	if (self->cancel_state == PTHREAD_CANCEL_DISABLE) {
-		self->cancel_state = PTHREAD_CANCEL_ENABLE;
+	self = to_posix_thread(pthread_self());
+	if (self == NULL) {
+		/* not a valid posix_thread */
+		__ASSERT_NO_MSG(self != NULL);
+		k_thread_abort(k_current_get());
 	}
 
-	k_spin_unlock(&self->cancel_lock, cancel_key);
-
-	pthread_mutex_lock(&self->state_lock);
-	if (self->state == PTHREAD_JOINABLE) {
-		self->state = PTHREAD_EXITED;
-		self->retval = retval;
-		pthread_cond_broadcast(&self->state_cond);
-	} else {
-		self->state = PTHREAD_TERMINATED;
-	}
+	/* Make a thread as cancelable before exiting */
```

Comment:
Ok, this looks right -- if `pthread_cancel()` is called on this thread while the destructors are being processed, then we'll never get to the `posix_thread_finalize()` call below. Do we want to mark the thread as not cancelable before calling `posix_thread_finalize()` so that process doesn't get interrupted mid-way?

Reply:
Sounds reasonable

Reply:
I put the destructor calls into `posix_thread_finalize()` and the passed the thread to terminate to that function instead of `pthread_self()`.